[BE] add test selector to test_testing

walterddr opened the pull request:
This is a reflection of recent failures in #55753 and #55522. 
We are lacking a test to safeguard these test env var.

Test Plan
1. CI
2. Run locally using `python test/test_testing.py -k test_filtering_env_var -v`
  - gives failure on 2ca45cb9e8 and d0cd16899f
  - passes on 159e1100bf and current master

malfet commented on test/test_testing.py:
```diff
@@ -682,6 +683,56 @@ def test_calculate_2_shards_against_optimal_shards(self):
+"""
+        # Test without setting env var
+        env = dict(os.environ)
+        if 'PYTORCH_TESTING_DEVICE_ONLY_FOR' in env.keys():
+            del env['PYTORCH_TESTING_DEVICE_ONLY_FOR']
+        if 'PYTORCH_TESTING_DEVICE_EXCEPT_FOR' in env.keys():
+            del env['PYTORCH_TESTING_DEVICE_EXCEPT_FOR']
```
Nit
```suggestion
       for key in ['PYTORCH_TESTING_DEVICE_ONLY_FOR', 'PYTORCH_TESTING_DEVICE_EXCEPT_FOR']:
          if key in env.keys():
             del env[key]
```

malfet reviewed (approved):
In order to avoid typos/rename regressions, can you please instead of duplicating string literals define `PYTORCH_TESTING_DEVICE_ONLY_FOR='PYTORCH_TESTING_DEVICE_ONLY_FOR'` somewhere in `torch.testing._internal` and use those constants both in filter as well as in this test?

walterddr commented on test/test_testing.py:
```diff
@@ -682,6 +685,60 @@ def test_calculate_2_shards_against_optimal_shards(self):
                 # All the tests should be represented by some shard
                 self.assertEqual(sorted_tests, sorted_shard_tests)
 
+    @@unittest.skipIf(IS_WINDOWS, "Skipping because doesn't work for windows")
```
```suggestion
    @unittest.skipIf(IS_WINDOWS, "Skipping because doesn't work for windows")
```

Commit:
[BE] add test selector to test_testing (#55931)

Summary:
This is a reflection of recent failures in https://github.com/pytorch/pytorch/issues/55753 and https://github.com/pytorch/pytorch/issues/55522.
We are lacking a test to safeguard these test env var.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/55931

Test Plan:
1. CI
2. Run locally using `python test/test_testing.py -k test_filtering_env_var -v`
  - gives failure on 2ca45cb9e8 and d0cd16899f
  - passes on 159e1100bf and current master

Reviewed By: jbschlosser

Differential Revision: D27747537

Pulled By: walterddr

fbshipit-source-id: 9b70470f1cb8cb6440f930e7e52119aed5374dfc

jeffdaily commented:
@walterddr rocm never had a clean CI signal before landing.  This is now causing failures on master for rocm.

VitalyFedyunin commented:
Fails ROCm tests https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.1-py3.6-test2/2152//console

walterddr commented:
this is fixed in #56178
